docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,346 @@
+and define different struct to implement this interface.
+for `Pod` resource, we can define 
+```
+const (
+	ResourceTypePod        = "pod"
+)
+type PodResource struct {
+	ResourceType string
+	NameSpace    string
+	NodeID       string
+	ID           string
+}
+
+func (this *PodResource) GetResourceType() string {
+	return this.ResourceType
+}
+func (this *PodResource) GetNameSpace() string {
+	return this.NameSpace
+}
+func (this *PodResource) GetNodeID() string {
+	return this.NodeID
+}
+func (this *PodResource) GetID() string {
+	return this.ID
+}
+```
+
+for `ConfigMap` resource, we can define:
+```
+const (
+	ResourceTypeConfigmap  = "configmap"
+)	
+type ConfigMapResource struct {
+	ResourceType string
+	NameSpace    string
+	NodeID       string
+	ID           string
+}
+
+func (this *ConfigMapResource) GetResourceType() string {
+	return this.ResourceType
+}
+func (this *ConfigMapResource) GetNameSpace() string {
+	return this.NameSpace
+}
+func (this *ConfigMapResource) GetNodeID() string {
+	return this.NodeID
+}
+func (this *ConfigMapResource) GetID() string {
+	return this.ID
+}
+```
+
+those types must registerd by `RegisterType()` function. then we can convert those type into `Any` type.
+```
+	func init() {
+    	var _ ToEdgeResourcer = &PodResource{}
+    	var _ ToEdgeResourcer = &ConfigMapResource{}
+    
+    	typeurl.RegisterType(PodResource{}, "pod")
+    	typeurl.RegisterType(ConfigMapResource{}, "configMap")
+    }
+```
+
+take `syncPod()`(cloud/pkg/controller/controller/downstream.go) function as a example:
+
+currently, it build Resource by `messagelayer.BuildResource(()`
+
+```
+			msg := model.NewMessage("")
+			resource, err := messagelayer.BuildResource(pod.Spec.NodeName, pod.Namespace, model.ResourceTypePod, pod.Name)
+			if err != nil {
+				log.LOGGER.Warnf("built message resource failed with error: %s", err)
+				continue
+			}
+```
+
+we can change it to 
+```
+    resource := &PodResource{
+        ID:        pod.Name,
+        NodeID:    pod.Spec.NodeName,
+        NameSpace: pod.Namespace,
+        ResourceType: ResourceTypePod,
+    }
+```
+
+in cloudhub model (`cloud/pkg/cloudhub/channelq/channelq.go`), it will receive this message:
+```
+func (q *ChannelEventQueue) dispatchMessage() {
+	for {
+		msg, err := q.ctx.Receive("cloudhub")
+		if err != nil {
+			log.LOGGER.Infof("receive not Message format message")
+			continue
+		}
+		resource := msg.Router.Resource
+		tokens := strings.Split(resource, "/")
+		numOfTokens := len(tokens)
+```
+
+this main function is to get the hostid , and put the message on the hostid\`s channel. but it use `strings.Split()` function to get hostid
+
+we can change to :
+```
+    res := msg.GetResource()
+
+	teRes, ok := res.(resources.ToEdgeResourcer)
+	if !ok || teRes.GetNodeID() == "" {
+	    continue
+	}
+	nodeID := teRes.GetNodeID()
+
+```
+and finnally, we convert Message to Event :
+```
+event := model.MessageToEvent(&msg)
+		select {
+		case rChannel <- event:
+		}
+```
+
```

Comment:
i think `Event` struct is useless.   Shall we consider using `Message` directly ??